File: problems/brace_expansion/brace_expansion_test.cc
#include "brace_expansion.h"

#include <gtest/gtest.h>
#include <string>
#include <vector>

struct BraceExpansionCase {
    const std::string test_name;
    const std::string input;
    const std::vector<std::string> expected;
};

using BraceExpansionTest = ::testing::TestWithParam<BraceExpansionCase>;

TEST_P(BraceExpansionTest, TestCases) {
    const BraceExpansionCase &testCase = GetParam();
    const auto result = braceExpansion(testCase.input);
    EXPECT_EQ(result, testCase.expected);
}

INSTANTIATE_TEST_SUITE_P(BraceExpansionTestCases, BraceExpansionTest,
                         ::testing::Values(
                             BraceExpansionCase{
                                 .test_name = "BasicCase",
                                 .input = "{a,b}c{d,e}f",
                                 .expected = {"acdf", "acef", "bcdf", "bcef"},
                             },
                             BraceExpansionCase{
                                 .test_name = "NoBraces",
                                 .input = "abcd",
                                 .expected = {"abcd"},
                             },
                             BraceExpansionCase{
                                 .test_name = "LeadingLiteralFollowedByBraces",
                                 .input = "k{a,b,c,d,e,f,g}",
                                 .expected = {"ka", "kb", "kc", "kd", "ke", "kf", "kg"},
                             }),
                         [](const testing::TestParamInfo<BraceExpansionCase> &info) {
                             return info.param.test_name;
                         });

File: problems/brace_expansion/brace_expansion_test.py
"""Test cases for the brace_expansion function."""

import pytest

from brace_expansion import braceExpansion


@pytest.mark.parametrize(
    "s, expected",
    [
        ("{a,b}c{d,e}f", ["acdf", "acef", "bcdf", "bcef"]),  # Basic Case
        ("abcd", ["abcd"]),  # No Braces
        (
            "k{a,b,c,d,e,f,g}",
            ["ka", "kb", "kc", "kd", "ke", "kf", "kg"],
        ),  # Leading Literal followed by Braces
    ],
    ids=[
        "Basic Case",
        "No Braces",
        "Leading Literal followed by Braces",
    ],
)
def test_brace_expansion(s, expected):
    assert braceExpansion(s) == expected
